controller/lib/hal/hal_stm32.cpp

```diff
@@ -0,0 +1,856 @@
+
+      for( uint16_t i=0; i<len; i++ ) {
+         int ch = rxDat.Get();
+         if( ch < 0 ) return i;
+         *buf++ = ch;
+      }
+      return len;
+   }
+
+   uint16_t write( const uint8_t *buf, uint16_t len) {
+
+      uint16_t i;
+      for( i=0; i<len; i++ ){
+         if( !txDat.Put( *buf++ ) )
+            break;
+      }
+
+      // Enable the tx interrupt
+      reg->ctrl[0] |= 0x0080;
+
+      return i;
+   }
+
+   int RxFull( void ){
+      return rxDat.FullCt();
+   }
+
+   int TxFree( void ){
+      return txDat.FreeCt();
+   }
+};
+
+static UART rpUART;
+
+// The UART that talks to the rPi uses the following pins:
+//    PB10 - TX
+//    PB11 - RX
+//    PB13 - RTS
+//    PB14 - CTS
+//
+// These pins are connected to UART3
+static void InitUARTs( void )
+{
+// NOTE - The UART functionality hasn't been tested due to lack of hardware!
+//        Need to do that as soon as the boards are available.
+   EnableClock( UART3_BASE );
+
+   GPIO_PinAltFunc( GPIO_B_BASE, 10, 7 );
+   GPIO_PinAltFunc( GPIO_B_BASE, 11, 7 );
+   GPIO_PinAltFunc( GPIO_B_BASE, 13, 7 );
+   GPIO_PinAltFunc( GPIO_B_BASE, 14, 7 );
+
+   rpUART.Init( (UART_Regs*)UART3_BASE, 115200 );
+
+   EnableInterrupt( INT_VEC_UART3, 3 );
+}
+
+static void UART3_ISR( void ){
+   rpUART.ISR();
+}
+
+uint16_t HalApi::serialRead(char *buf, uint16_t len) {
+   return rpUART.read( (uint8_t *)buf, len );
+}
+
+uint16_t HalApi::serialBytesAvailableForRead() {
+   return rpUART.RxFull();
+}
+
+uint16_t HalApi::serialWrite(const char *buf, uint16_t len) {
+   return rpUART.write( (const uint8_t *)buf, len);
+  return 0;
+}
+
+uint16_t HalApi::serialBytesAvailableForWrite() {
+   return rpUART.TxFree();
+}
+
+/******************************************************************
+ * Watchdog timer
+ *****************************************************************/
+void HalApi::watchdog_init() {
+}
+
+void HalApi::watchdog_handler() {
+}
+
+// Enable clocks to a specific peripherial.
+// On the STM32 the clocks going to various peripherials on the chip
+// are individually selectable and for the most part disabled on startup.
+// Clocks to the specific peripherials need to be enabled through the 
+// RCC (Reset and Clock Controller) module before the peripherial can be
+// used.
+// Pass in the base address of the peripherial to enable it's clock
+static void EnableClock( uint32_t base )
+{
+   // I don't include all the peripherials here, just the ones
+   // that we currently use or seem likely to be used in the 
+   // future.  To add more peripherials, just look up the appropriate
+   // bit in the reference manual RCC chapter.
+   //
+   // This big case statement finds the index of the register in the 
+   // array of clock enable registers, and the bit number used to enable
+   // the clock for the specified peripherial.
+   int ndx = -1;
+   int bit = 0;
+   switch( base )
+   {
+      case DMA1_BASE:    ndx = 0; bit =  0; break;
```

Comment:
In this case and in case of the other similar tables, how can we be certain that they don't contain a typo?

Reply:
The peripheral won't work if we don't enable it's clock, so I guess that in this case even the most minimal testing of the hardware will catch the problem.
For low level code like this I think you really need to manually test it to ensure that it's actually working.  That goes for this and any of the values being written to registers in this file.
Other then creating some type of virtual STM32 processor I don't really think there's a practical way to unit test the low level HAL code.